File: python/mldsa/field.py
"""Finite field arithmetic modulo Q = 8380417 for ML-DSA (FIPS 204)."""

Q = 8380417


def mod_q(a: int) -> int:
    """Reduce a modulo Q into range [0, Q)."""
    return a % Q


def field_add(a: int, b: int) -> int:
    """Add two field elements modulo Q."""
    return (a + b) % Q


def field_sub(a: int, b: int) -> int:
    """Subtract two field elements modulo Q."""
    return (a - b) % Q


def field_mul(a: int, b: int) -> int:
    """Multiply two field elements modulo Q."""
    return (a * b) % Q


def field_pow(base: int, exp: int) -> int:
    """Raise a field element to a power modulo Q."""
    return pow(base, exp, Q)


def field_inv(a: int) -> int:
    """Compute the modular inverse of a modulo Q using Fermat's little theorem."""
    if a % Q == 0:
        raise ValueError("Cannot invert zero")
    return pow(a, Q - 2, Q)

File: go/internal/mldsa/field.go
// Package mldsa implements internal functions for ML-DSA (FIPS 204).
package mldsa

// Q is the ML-DSA prime modulus.
const Q = 8380417

// ModQ reduces a (possibly negative) int64 to [0, Q).
func ModQ(a int64) int {
	r := int(a % int64(Q))
	if r < 0 {
		r += Q
	}
	return r
}

// FieldAdd returns (a + b) mod Q with inputs in [0, Q).
func FieldAdd(a, b int) int {
	r := a + b
	if r >= Q {
		r -= Q
	}
	return r
}

// FieldSub returns (a - b) mod Q with inputs in [0, Q).
func FieldSub(a, b int) int {
	r := a - b
	if r < 0 {
		r += Q
	}
	return r
}

// FieldMul returns (a * b) mod Q. Uses int64 intermediates to avoid overflow.
func FieldMul(a, b int) int {
	return int(int64(a) * int64(b) % int64(Q))
}

// FieldPow returns a^e mod Q by repeated squaring.
func FieldPow(a, e int) int {
	result := int64(1)
	base := int64(a) % int64(Q)
	exp := e
	for exp > 0 {
		if exp&1 == 1 {
			result = result * base % int64(Q)
		}
		exp >>= 1
		base = base * base % int64(Q)
	}
	return int(result)
}

// FieldInv returns the modular inverse of a mod Q using Fermat's little theorem.
func FieldInv(a int) int {
	return FieldPow(a, Q-2)
}
